### ISSUE TITLE
unsortedOfflinePrimaryVertices4D speedup for 2023 timing workflows

### ISSUE THREAD

**Issue by slava77** [cms-sw/cmssw#17826]
related to discussion in #17610 leading to 
https://github.com/cms-sw/cmssw/pull/17610#issuecomment-285035770

**Comment by slava77** [comment]
assign upgrade,reconstruction

@lgray

**Comment by lgray** [comment]
Thanks. I'll try to get to it soon.

**Comment by Dr15Jones** [comment]
I do not believe vectorization will help this algorithm that much. vectorization could, at most, speed the algorithm up by a factor of 8, but from the measurements we were doing on the KNL the algorithm can be a 100x to 1000x slower than all other modules in the job. If you start with an event with very many primary vertices we saw it can takes 4 hours for the algorithm to converge.

**Comment by lgray** [comment]
Hi Chris, right now we are also dealing with an increase in the bulk timing
(not just these outlier events, which we hope to eventually deal with).
Vectorization is important here, and from the vectorization of the 1D
version of the algorithm they gained a factor 2 or so, maybe we end up
doing better. Either way the average time of the vertexing needs to be
reduced at this point.

Investigating the very long tails will come too.

-L

**Comment by slava77** [comment]
@lgray 
please check if you also need this fix 
https://github.com/cms-sw/cmssw/pull/16764/files#diff-433c2174a94e2712d80a0b75bba3a49cL196
around https://github.com/lgray/cmssw/blob/8929924570c2ad59f9d21762df80786ed141e4d5/RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZT.cc#L110

```diff
    // normalization
-    if (tks[i].zi>0){
+    if (tks[i].zi>epsilon){//e.g. constexpr double epsilon = 1.e-100
```

**Comment by slava77** [comment]
@lgray 
what is the status of this issue?
Please "@" others who may be working on this.
Thank you.

**Comment by kpedro88** [comment]
+1
(fully) resolved by #20709
